tlv/primitive.go

```diff
@@ -0,0 +1,272 @@
+		return nil
+	}
+	return ErrTypeForDecoding{val, "uint16", l, 2}
+}
+
+// DUint32 is a Decoder for uint32 values. An error is returned if val is not a
+// *uint32.
+func DUint32(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
+	if i, ok := val.(*uint32); ok && l == 4 {
+		if _, err := io.ReadFull(r, buf[:4]); err != nil {
+			return err
+		}
+		*i = byteOrder.Uint32(buf[:4])
+		return nil
+	}
+	return ErrTypeForDecoding{val, "uint32", l, 4}
+}
+
+// DUint64 is a Decoder for uint64 values. An error is returned if val is not a
+// *uint64.
+func DUint64(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
+	if i, ok := val.(*uint64); ok && l == 8 {
+		if _, err := io.ReadFull(r, buf[:]); err != nil {
+			return err
+		}
+		*i = byteOrder.Uint64(buf[:])
+		return nil
+	}
+	return ErrTypeForDecoding{val, "uint64", l, 8}
+}
+
+// EBytes32 is an Encoder for 32-byte arrays. An error is returned if val is not
+// a *[32]byte.
+func EBytes32(w io.Writer, val interface{}, _ *[8]byte) error {
+	if b, ok := val.(*[32]byte); ok {
+		_, err := w.Write(b[:])
+		return err
+	}
+	return ErrTypeForEncoding{val, "[32]byte"}
+}
+
+// DBytes32 is a Decoder for 32-byte arrays. An error is returned if val is not
+// a *[32]byte.
+func DBytes32(r io.Reader, val interface{}, _ *[8]byte, l uint64) error {
+	if b, ok := val.(*[32]byte); ok && l == 32 {
+		_, err := io.ReadFull(r, b[:])
+		return err
+	}
+	return ErrTypeForDecoding{val, "[32]byte", l, 32}
+}
+
+// EBytes33 is an Encoder for 33-byte arrays. An error is returned if val is not
+// a *[33]byte.
+func EBytes33(w io.Writer, val interface{}, _ *[8]byte) error {
+	if b, ok := val.(*[33]byte); ok {
+		_, err := w.Write(b[:])
+		return err
+	}
+	return ErrTypeForEncoding{val, "[33]byte"}
+}
+
+// DBytes33 is a Decoder for 33-byte arrays. An error is returned if val is not
+// a *[33]byte.
+func DBytes33(r io.Reader, val interface{}, _ *[8]byte, l uint64) error {
+	if b, ok := val.(*[33]byte); ok {
+		_, err := io.ReadFull(r, b[:])
+		return err
+	}
+	return ErrTypeForDecoding{val, "[33]byte", l, 33}
+}
+
+// EBytes64 is an Encoder for 64-byte arrays. An error is returned if val is not
+// a *[64]byte.
+func EBytes64(w io.Writer, val interface{}, _ *[8]byte) error {
+	if b, ok := val.(*[64]byte); ok {
+		_, err := w.Write(b[:])
+		return err
+	}
+	return ErrTypeForEncoding{val, "[64]byte"}
+}
+
+// DBytes64 is an Decoder for 64-byte arrays. An error is returned if val is not
+// a *[64]byte.
+func DBytes64(r io.Reader, val interface{}, _ *[8]byte, l uint64) error {
+	if b, ok := val.(*[64]byte); ok && l == 64 {
+		_, err := io.ReadFull(r, b[:])
+		return err
+	}
+	return ErrTypeForDecoding{val, "[64]byte", l, 64}
+}
+
+// EPubKey is an Encoder for *btcec.PublicKey values. An error is returned if
+// val is not a **btcec.PublicKey.
+func EPubKey(w io.Writer, val interface{}, _ *[8]byte) error {
+	if pk, ok := val.(**btcec.PublicKey); ok {
+		_, err := w.Write((*pk).SerializeCompressed())
```

Comment:
Do we really need methods for `PublicKey`, or could we just require the caller to use `[33]byte`?

Reply:
the test vectors in the spec require us to fail when reading an invalid point, so only parsing `[33]byte` isn't strict enough. arguably that check could be deferred to a higher level, but it this is the safest approach